api/desecapi/models.py

```diff
@@ -216,80 +216,87 @@ def write_rrsets(self, datas):
-        RRset.objects.filter(q_delete, domain=self).delete()
-
-        # Prepare and save new RRset contents
-        # We modify the rrsets dictionary and thus loop over a copy of it
-        for rrset, rrs in list(rrsets.items()):
-            if not rrs:
-                continue
-            # (Create and) get correct RRset and update dictionary accordingly
-            del rrsets[rrset]
-            (rrset, _) = RRset.objects.get_or_create(domain=self,
-                                                     subname=rrset.subname,
-                                                     type=rrset.type,
-                                                     ttl=rrset.ttl)
-            rrsets[rrset] = [RR(rrset=rrset, content=rr.content) for rr in rrs]
-
-        RR.objects.bulk_create([rr for rrs in rrsets.values() for rr in rrs])
-
-        # Send changed RRsets to pdns
-        if rrsets and not self.owner.locked:
-            pdns.set_rrsets(self, rrsets)
+                rrsets_to_write[rrset] = []
+                q_empty |= q
+
+        # Construct querysets representing RRsets that do (not) have RR
+        # contents and lock them
+        qs_meaty = rrset_qs.filter(q_meaty).select_for_update()
+        qs_empty = rrset_qs.filter(q_empty).select_for_update()
+
+        # For existing RRsets, execute TTL updates and mark for RR update
+        rrsets_same_rrs = []
+        for rrset in qs_meaty.all():
+            rrset_temp = rrsets_meaty_todo.pop((rrset.subname, rrset.type))
+            rrs_temp = {rr.content for rr in rrsets[rrset_temp]}
+            rrs = {rr.content for rr in rrset.records.all()}
+
+            changed_ttl = (rrset_temp.ttl != rrset.ttl)
+            changed_rrs = (rrs_temp != rrs)
+
+            if changed_ttl:
+                rrset.ttl = rrset_temp.ttl
+                rrset.save()
+            if changed_ttl or changed_rrs:
+                rrsets_to_write[rrset] = [RR(rrset=rrset, content=rr_content)
+                                          for rr_content in rrs_temp]
+            if not changed_rrs:
+                rrsets_same_rrs.append(rrset)
+
+        # At this point, rrsets_meaty_todo points to new RRsets only. Let's
+        # save them. This does not add the RR contents.
+        for key, rrset in list(rrsets_meaty_todo.items()):
+            rrset.save()
+            rrsets_to_write[rrset] = rrsets[rrset]
+
+        # Repeat lock to make sure new RRsets are also locked
+        rrset_qs.filter(q_meaty).select_for_update()
+
+        # Delete empty RRsets
+        qs_empty.delete()
+
+        # Update contents of modified RRsets
+        RR.objects.filter(rrset__in=qs_meaty).exclude(rrset__in=rrsets_same_rrs).delete()
+        RR.objects.bulk_create([rr
+                                for (rrset, rrs) in rrsets_to_write.items()
+                                for rr in rrs
+                                if rrset not in rrsets_same_rrs])
```

Comment:
is there any advantage in actually updating django objects instead of deleting and re-creating them with new content?

Reply:
You mean, in updating the `RR` objects? It could be done, but their number could change (say, someone adds a second IP address to an RRset of type `A`). We would then have to figure out 1) in case of increase, how many can be updated and then create the other ones only, 2) in case of decrease how many to update and how many to delete.

Since the deletion and the creation actually are bulk SQL statements, I think that this approach is not only much easier and less error prone, but also potentially more efficient.

Reply:
... not to mention wiring it all up correctly with the parent `RRset`